pypi-setuptools.yaml

```
@@ -0,0 +1,60 @@
+package:
+  name: pypi-setuptools
+  version: 69.2.0
+  epoch: 1
+  description: Easily download, build, install, upgrade, and uninstall Python packages
+  copyright:
+    - license: "MIT"
+
+environment:
+  contents:
+    packages:
+      - build-base
+      - busybox
+      - ca-certificates-bundle
+      - python-3.10-base
+      - python-3.11-base
+      - python-3.12-base
+      - wolfi-base
+  environment:
+    # This is needed to work around the error "ValueError: ZIP does not support timestamps before 1980"
+    SOURCE_DATE_EPOCH: 315532800
+
+vars:
+  pypi-package: setuptools
+
+data:
+  - name: py-versions
+    items:
+      3.12: ""
+      3.11: ""
+      3.10: ""
+
+pipeline:
+  - uses: fetch
+    with:
+      expected-sha256: 0ff4183f8f42cd8fa3acea16c45205521a4ef28f73c6391d8a25e92893134f2e
+      uri: https://files.pythonhosted.org/packages/source/s/setuptools/setuptools-${{package.version}}.tar.gz
+
+  - runs: mkdir -p ${{targets.destdir}}
+
+subpackages:
+  - range: py-versions
+    name: py${{range.key}}-${{vars.pypi-package}}
+    description: Easily download, build, install, upgrade, and uninstall Python packages
+    pipeline:
+      - runs: |
+          python${{range.key}} setup.py install --no-compile --root="${{targets.subpkgdir}}" --prefix=/usr
+          rm -rf build/
+          python${{range.key}} -m compileall --invalidation-mode=unchecked-hash -r100 ${{targets.subpkgdir}}/usr/lib/
+
+      - uses: strip
+    dependencies:
+      provides:
+        - py3-setuptools=${{package.full-version}}-${{range.key}}
```

Is this version apk version compliance?
Can no be
```suggestion
        - py${{range.key}}-setuptools=${{package.full-version}}
```

---

@mesaglio ,

I didn't follow the question "Is this version apk compliance?"

I think the intent of 'py3-setuptools'  provides is to allow something to depend on "latest setuptools", similar to the 'python3' provides that we've had before. The newest available version would be the highest.

I think your suggested provides is (py3.12-setuptools) ends up being the same as the package name, right?

---

The variabel `package.full-version` have, for example the follow format `1.36.0-r6` (**version**-**epoch**) value. We need to check if we define, a version formated **version**-**epoch**-**py_version**, dont break another thing. For example, https://github.com/chainguard-dev/melange/blob/165db39be30913d0a1c98258ba4e1451d3002d45/pkg/config/config.go#L1020

---

@mesaglio i see what you mean, and I do agree mostly. We don't need the `py${{range.key}}-setuptools=${{package.full-version}}` as that's recursive and is already the name of this subpackage. We do need to provide `py3-setuptools` somewhere, as all previous builds did that I think (for backwards compat). Or maybe we need to give up on that, if we want to allow all setuptools to be co-installable for multiple python versions. As provides conflict with each other right?

---

Ideally i wish them to have provides at different priorities.

But somehow failing to make that work.

---

making py3-setuptools an explicit meta-subpackage.